### PR TITLE
Abstract storage backend

### PR DESCRIPTION
**Pull request by cannium**
Add an interface storage `Cluster`, for unit tests and new backends. Code is modified from part of #194

### CODE REVIEW

File: backend/backend.go [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+}
+
+// Backend plugins should implement this interface
+type Plugin interface {
+	// initialize backend cluster handlers,
+	// returns cluster ID -> Cluster, panic on errors
+	Initialize(logger *log.Logger, config helper.Config) map[string]Cluster
+	// pick a cluster for specific object upload
+	// XXX: this is ugly and subject to change
+	PickCluster(clusters map[string]Cluster, weights map[string]int,
```
Comment (cxt90730) [inline backend/backend.go]: picking a pool may be irrelevant with backend type

Comment (cannium) [reply backend/backend.go]: Ideally so, but for some reason you cannot write a ceph cluster to its full capacity...

File: ceph/cluster.go [hunk 2]
```diff
@@ -124,27 +145,27 @@ func drain_pending(p *list.List) int {
 	return ret
 }
 
-func (cluster *CephStorage) GetUniqUploadName() string {
+func (cluster *CephCluster) getUniqUploadName() string {
 	cluster.CountMutex.Lock()
 	defer cluster.CountMutex.Unlock()
```
Comment (chenji-kael) [inline ceph/cluster.go]: lock less could be better

**Comment by cxt90730** [comment]
resolve conflicts first...

**Pull request merged**